[LV] Add support for cmp reductions with decreasing IVs.

fhahn opened the pull request:
Similar to FindLastIV, add FindFirstIV to support select (icmp(), x, y)
reductions where one of x or y is a decreasing induction. This is done
via two new recurrence kinds FindFirstIV(UMin|SMin), which selects the first
 value from the reduction vector using umin/smin instead of the last value
(FindLastIV). It uses unsigned max/signed max as sentinel value. The
SMin version is used if the IV range does not include signed max.
Otherwise UMin is used if the IV range does not include signed max.

Alternatively we could consolidate FindFirstIV/FindLastIV in a single
FindIV kind, and add a field indicating if whether we want to select the
first or last value.

artagnon commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -259,17 +267,33 @@ class RecurrenceDescriptor {
+  /// Returns true if the recurrence kind is of the form
+  ///   select(cmp(),x,y) where one of (x,y) is an increasing or decreasing loop
+  ///   induction.
+  static bool isFindIVRecurrenceKind(RecurKind Kind) {
```
Factor out isFindFirstIVRecurrenceKind?

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -736,11 +757,18 @@ RecurrenceDescriptor::isFindLastIVPattern(Loop *TheLoop, PHINode *OrigPhi,
   // We are looking for selects of the form:
   //   select(cmp(), phi, increasing_loop_induction) or
   //   select(cmp(), increasing_loop_induction, phi)
-  // TODO: Support for monotonically decreasing induction variable
-  if (!IsIncreasingLoopInduction(NonRdxPhi))
+  if (Kind == RecurKind::FindLastIV) {
+    if (IsSupportedLoopInduction(NonRdxPhi, Kind))
+      return InstDesc(I, Kind);
     return InstDesc(false, I);
+  }
+
+  if (IsSupportedLoopInduction(NonRdxPhi, RecurKind::FindFirstIVSMin))
+    return InstDesc(I, RecurKind::FindFirstIVSMin);
+  if (IsSupportedLoopInduction(NonRdxPhi, RecurKind::FindFirstIVUMin))
+    return InstDesc(I, RecurKind::FindFirstIVUMin);
 
-  return InstDesc(I, RecurKind::FindLastIV);
+  return InstDesc(false, I);
```
Get IsSupportedLoopInduction to return a Kind, and compare with `Kind ==` to unify this code?

artagnon commented on llvm/test/Transforms/LoopVectorize/if-reduction.ll:
```diff
@@ -1916,4 +1978,8 @@ for.body:                                         ; preds = %entry, %for.body
 ; CHECK: [[LOOP29]] = distinct !{[[LOOP29]], [[META2]], [[META1]]}
 ; CHECK: [[LOOP30]] = distinct !{[[LOOP30]], [[META1]], [[META2]]}
 ; CHECK: [[LOOP31]] = distinct !{[[LOOP31]], [[META2]], [[META1]]}
+; CHECK: [[LOOP32]] = distinct !{[[LOOP32]], [[META1]], [[META2]]}
```
--check-globals none?

artagnon commented on llvm/test/Transforms/LoopVectorize/if-reduction.ll:
Move the changed tests into iv-select-cmp?

artagnon commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -23325,6 +23329,8 @@ class HorizontalReduction {
     case RecurKind::FMulAdd:
     case RecurKind::AnyOf:
     case RecurKind::FindLastIV:
+    case RecurKind::FindFirstIVUMin:
+    case RecurKind::FindFirstIVSMin:
```
Why is this asymmetric? Wouldn't it be nicer to first separate out FindLastIV into FindLastIVSMax/UMax?

artagnon commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1244,12 +1244,16 @@ Value *llvm::createAnyOfReduction(IRBuilderBase &Builder, Value *Src,
   Value *Sentinel = Desc.getSentinelValue();
   Value *MaxRdx = Src->getType()->isVectorTy()
-                      ? Builder.CreateIntMaxReduce(Src, true)
+                      ? (Desc.getRecurrenceKind() == RecurKind::FindLastIV
+                             ? Builder.CreateIntMaxReduce(Src, true)
+                             : Builder.CreateIntMinReduce(
+                                   Src, Desc.getRecurrenceKind() ==
+                                            RecurKind::FindFirstIVSMin))
```
This ternary expression reads badly?

artagnon commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -259,17 +267,33 @@ class RecurrenceDescriptor {
+  /// Returns the sentinel value for FindFirstIV &FindLastIV recurrences to
+  /// replace the start value.
   Value *getSentinelValue() const {
-    assert(isFindLastIVRecurrenceKind(Kind) && "Unexpected recurrence kind");
```
Why strip this assert, and move it to a branch?

artagnon commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -56,6 +56,13 @@ enum class RecurKind {
   FindLastIV, ///< FindLast reduction with select(cmp(),x,y) where one of
               ///< (x,y) is increasing loop induction, and both x and y are
               ///< integer type.
+  FindFirstIVUMin, /// FindFirst reduction with select(icmp(),x,y) where one of
+                   ///< (x,y) is a decreasing loop induction, and both x and y
+                   ///< are integer type.
+  FindFirstIVSMin /// FindFirst reduction with select(icmp(),x,y) where one of
+                  ///< (x,y) is a decreasing loop induction, and both x and y
+                  ///< are integer type.
```
Identical doc?

Mel-Chen replied:
This patch introduces too many new idioms at once.
To keep the patch reviewable and focused, I suggest we pick one direction to extend first:
either support unsigned FindLastIV, or add signed FindFirstIV support.

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -710,21 +713,39 @@ RecurrenceDescriptor::isFindLastIVPattern(Loop *TheLoop, PHINode *OrigPhi,
     //   [<sentinel value> + 1, <sentinel value>)
-    // where <sentinel value> is SignedMin(<recurrence type>)
+    // where <sentinel value> is SignedMin(<recurrence type>) for FindLast or
+    // UnsignedMax(<recurrence type>) for FindFirst.
```
We also have a case with SignedMax.

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -710,21 +713,39 @@ RecurrenceDescriptor::isFindLastIVPattern(Loop *TheLoop, PHINode *OrigPhi,
+    } else if (!SE.isKnownPositive(Step))
       return false;
 
     const ConstantRange IVRange = SE.getSignedRange(AR);
```
Shouldn't this be getUnsignedRange in some cases?

artagnon replied:
Strip?

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -710,21 +713,39 @@ RecurrenceDescriptor::isFindLastIVPattern(Loop *TheLoop, PHINode *OrigPhi,
+                                 : APInt::getSignedMinValue(NumBits);
+      ValidRange = ConstantRange::getNonEmpty(Sentinel + 1, Sentinel);
+    }
     LLVM_DEBUG(dbgs() << "LV: FindLastIV valid range is " << ValidRange
                       << ", and the signed range of " << *AR << " is "
                       << IVRange << "\n");
```
Outdated DBG message?

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -990,6 +1018,12 @@ bool RecurrenceDescriptor::isReductionPHI(PHINode *Phi, Loop *TheLoop,
     LLVM_DEBUG(dbgs() << "Found a FindLastIV reduction PHI." << *Phi << "\n");
     return true;
   }
+  if (AddReductionVar(Phi, RecurKind::FindFirstIVUMin, TheLoop, FMF, RedDes, DB,
+                      AC, DT, SE)) {
+    LLVM_DEBUG(dbgs() << "Found a FindFirstV reduction PHI." << *Phi << "\n");
+    return true;
+  }
```
I think it may be possible to use RecurrenceDescriptor::isSigned() to unify? Not sure.

artagnon replied:
See #141752. You can rebase on top of it, assuming it's a good direction.

Mel-Chen commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -259,17 +267,33 @@ class RecurrenceDescriptor {
+  /// Returns the sentinel value for FindFirstIV &FindLastIV recurrences to
+  /// replace the start value.
   Value *getSentinelValue() const {
-    assert(isFindLastIVRecurrenceKind(Kind) && "Unexpected recurrence kind");
     Type *Ty = StartValue->getType();
-    return ConstantInt::get(Ty,
-                            APInt::getSignedMinValue(Ty->getIntegerBitWidth()));
+    if (isFindLastIVRecurrenceKind(Kind)) {
+      return ConstantInt::get(
+          Ty, APInt::getSignedMinValue(Ty->getIntegerBitWidth()));
+    } else if (Kind == RecurKind::FindFirstIVSMin) {
+      return ConstantInt::get(
+          Ty, APInt::getSignedMaxValue(Ty->getIntegerBitWidth()));
+    } else {
+      assert(Kind == RecurKind::FindFirstIVUMin);
+      return ConstantInt::get(Ty, APInt::getMaxValue(Ty->getIntegerBitWidth()));
+    }
   }
```
https://github.com/llvm/llvm-project/pull/141788 Try to change the sentinel value to be selected dynamically instead of based on RecurKind statically.
If I'm not mistaken, the values for signed max, unsigned max, signed min, and unsigned min are all distinct.
So making the sentinel value non-static could help with supporting the unsigned version, I think.

fhahn committed:
[LV] Add support for cmp reductions with decreasing IVs using SMin.

Similar to FindLastIV, add FindFirstIV to support select (icmp(), x, y)
reductions where one of x or y is a decreasing induction. This is done
via a new recurrence kind FindFirstIVSMin, which selects the first
 value from the reduction vector using smin instead of the last value
(FindLastIV). It uses signed max as sentinel value. The

artagnon commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -269,22 +278,35 @@ class RecurrenceDescriptor {
 
-  /// Returns the sentinel value for FindLastIV recurrences to replace the start
-  /// value.
+  /// Returns the sentinel value for FindFirstIV &FindLastIV recurrences to
```
```suggestion
  /// Returns the sentinel value for FindFirstIV & FindLastIV recurrences to
```

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -741,9 +741,17 @@ Value *VPInstruction::generate(VPTransformState &State) {
+      MinMaxKind = IsSigned ? RecurKind::SMax : RecurKind::UMax;
+    } else {
+      assert(RecurrenceDescriptor::isFindFirstIVRecurrenceKind(RK) &&
+             "Kind must either be a FindLastIV or FindFirstIV");
```
```suggestion
             "Kind must either be FindLastIV or FindFirstIV");
```

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -741,9 +741,17 @@ Value *VPInstruction::generate(VPTransformState &State) {
+      assert(RecurrenceDescriptor::isFindFirstIVRecurrenceKind(RK) &&
+             "Kind must either be a FindLastIV or FindFirstIV");
+      assert(IsSigned &&
+             "only FindFirstIV with SMax is supported at the moment");
```
```suggestion
             "Only FindFirstIV with SMax is currently supported");
```

artagnon commented on llvm/test/Transforms/LoopVectorize/iv-select-cmp-decreasing.ll:
```diff
@@ -1,26 +1,185 @@
-; RUN: opt -passes=loop-vectorize -force-vector-interleave=1 -force-vector-width=4 -S < %s | FileCheck %s
-; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -S < %s | FileCheck %s
-; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=1 -S < %s | FileCheck %s
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=1 -force-vector-width=4 -S < %s | FileCheck --check-prefix=IC1VF4 %s
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -S < %s | FileCheck --check-prefix=IC4VF4 %s
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=1 -S < %s | FileCheck --check-prefix=IC4VF1 %s
```
```suggestion
; RUN: opt -passes=loop-vectorize -force-vector-interleave=1 -force-vector-width=4 -S < %s | FileCheck --check-prefix=CHECK,IC1VF4 %s
; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -S < %s | FileCheck --check-prefix=CHECK,IC4VF4 %s
; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=1 -S < %s | FileCheck --check-prefix=CHECK,IC4VF1 %s
```

fhahn committed:
!fixup address comments, thanks

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -713,36 +715,66 @@ RecurrenceDescriptor::isFindLastIVPattern(Loop *TheLoop, PHINode *OrigPhi,
     const SCEV *Step = AR->getStepRecurrence(SE);
-    if (!SE.isKnownPositive(Step))
+
+    if (isFindFirstIVRecurrenceKind(Kind)) {
+      if (!SE.isKnownNegative(Step))
+        return std::nullopt;
+    } else if (!SE.isKnownPositive(Step))
       return std::nullopt;
```
```suggestion
    if ((isFindFirstIVRecurrenceKind(Kind) && !SE.isKnownNegative(Step)) ||
        (isFindLastIVRecurrenceKind(Kind) && !SE.isKnownPositive(Step)))
        return std::nullopt;
```

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -713,36 +715,66 @@ RecurrenceDescriptor::isFindLastIVPattern(Loop *TheLoop, PHINode *OrigPhi,
+                                  : APInt::getMinValue(NumBits);
+        ValidRange = ConstantRange::getNonEmpty(Sentinel + 1, Sentinel);
+      } else {
+        assert(isFindFirstIVRecurrenceKind(Kind) &&
+               "Kind must either be FindLastIV or FindFirstIV");
```
```suggestion
```

There's already an assert on line 773.

artagnon commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1227,9 +1227,12 @@ Value *llvm::createFindLastIVReduction(IRBuilderBase &Builder, Value *Src,
-  Value *MaxRdx = Src->getType()->isVectorTy()
-                      ? Builder.CreateIntMaxReduce(Src, IsSigned)
-                      : Src;
+  Value *MaxRdx =
+      Src->getType()->isVectorTy()
+          ? (RecurrenceDescriptor::isFindLastIVRecurrenceKind(RdxKind)
+                 ? Builder.CreateIntMaxReduce(Src, IsSigned)
+                 : Builder.CreateIntMinReduce(Src, IsSigned))
+          : Src;
```
```suggestion
 Value *MinMaxRdx = RecurrenceDescriptor::isFindLastIVRecurrenceKind(RdxKind)
                 ? Builder.CreateIntMaxReduce(Src, IsSigned)
                 : Builder.CreateIntMinReduce(Src, IsSigned);
  Value *Rdx = Src->getType()->isVectorTy() ? MinMaxRdx : Src;
```

artagnon commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1324,8 +1327,8 @@ Value *llvm::createSimpleReduction(IRBuilderBase &Builder, Value *Src,
-         !RecurrenceDescriptor::isFindLastIVRecurrenceKind(Kind) &&
-         "AnyOf or FindLastIV reductions are not supported.");
+         !RecurrenceDescriptor::isFindIVRecurrenceKind(Kind) &&
+         "AnyOf, FindFirstIV and FindLastIV reductions are not supported.");
```
```suggestion
         "AnyOf and FindIV reductions are not supported.");
```

artagnon reviewed (approved):
LGTM, thanks! Some suggestions follow. Also, kindly update commit message before landing.

fhahn committed:
!fixup address latest comments, thanks!
